### PR TITLE
Display User ID for deleted users

### PR DESCRIPTION
**Pull request by HopperElec**
Closes issue https://github.com/mlomb/chat-analytics/issues/8

It might be useful to display the user ID of all users (probably only on hover, not inside the label as I've done here) but for now, to reduce data usage, I'm only storing the user ID for deleted users (and users with an avatar, since their ID was stored anyway).

Since the user ID used to be stored as a string for avatars, but it is now stored separately, the file size might actually be lower (assuming you have more users with avatars than deleted users).

I'm not good with design, so I have not applied any CSS to the ID, but it looks decent as it is.

### CODE REVIEW

**Comment by HopperElec** [comment]
I've just thought- most users are either deleted or have an avatar, so there's not much benefit to not storing the ID of users not matching the criteria and it might be reasonable to display the user ID in some form for all users. I'm still not completely sure on this, though, since mlomb is very determined to keep a low file size.

**Comment by mlomb** [comment]
Thanks for the contribution!

Storing the ID for every user can add up, and I prefer not to. If someone wants the ID of an user, they can search it in Discord I guess.

I don't like the idea of adding a new field to `IAuthors`, it hurt me when I had to add `da` already.
Also the `Label` component appears in a lot of places and we have to check if it look rights everywhere. For example we can't see the ID in the filter box.
image

I think this has a more simple solution, just set the name to something like "Deleted User #123456789", the UI is prepared for long names, we don't have to modify almost anything else. We might be able to remove the discriminator if it is a deleted user so we recover a bit of space.

**Comment by HopperElec** [comment]
> Storing the ID for every user can add up, and I prefer not to. If someone wants the ID of an user, they can search it in Discord I guess.
The reason I think storing the ID for every user might be useful is for if someone changes their username or discriminator after a report is generated, in which case they can't just "search it in Discord". I don't think it would add up too much because, as I said, it's stored for most users anyways

> I don't like the idea of adding a new field to `IAuthors`, it hurt me when I had to add `da` already.
What's the alternative? I'm trying to avoid modifying the structure of the code in any way for now because I'd imagine it would be difficult to pull in along with the other PRs

As for how to display it, yes I agree with replacing the discriminator. As I say, though, I'm terrible with design, so I'd prefer you choose exactly how to style it

**Comment by HopperElec** [comment]
> I don't like the idea of adding a new field to IAuthors, it hurt me when I had to add da already.
I don't use any of the other supported platforms enough to contribute for them, but other platforms have profile pictures/avatars and I'd imagine they can also have deleted users who still have IDs, so could it maybe be generalized to be useful for other platforms too?

File: report/components/core/Labels.tsx [hunk 1]
```diff
@@ -84,6 +84,9 @@ const _AuthorLabel = ({ index }: LabelProps) => {
     } else if (platform === "telegram") {
         // TODO: two letters
         let letter: string = "";
+        if (author.n === undefined) {
+            throw new Error("Telegram author name undefined (only Discord name's can be undefined)");
```
Comment (mlomb) [inline report/components/core/Labels.tsx:88]: This is very dangerous, throwing an error in something we don't know for sure it is true may crash the app unnecessarily.

**Comment by mlomb** [comment]
So here is the result
image

A note about storing IDs, I MAY change my mind but for now I prefer not to

File: report/components/core/Labels.tsx (hunk 1 above)
Comment (HopperElec) [reply report/components/core/Labels.tsx]: Before, author.n couldn't be undefined at all, so wouldn't an error have been thrown in this case anyways?

**Comment by HopperElec** [comment]
> I have no idea with Messenger, I don't have a deleted account in my messages.

I have someone in my Messenger from 4 years ago who deleted their account. I mention the age since Facebook may have changed how it is handled since, but I doubt it. I'll try exporting it later to see how it's represented in the export, but in the UI I can still see all messages and attachments, but I'm not able to send more messages, their profile picture is default (I can't remember if it used to be) and their name is missing

**Pull request merged**